### ISSUE TITLE
`dynamicParams=false` show 404 page in dev, but not show 404 page after build in prod

### ISSUE THREAD

**Issue by leozhao0709** [vercel/next.js#52765]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant Packages:
      next: 13.4.10-canary.8
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.9.5
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/wandering-forest-3tmmtz

### To Reproduce

Run `yarn dev`, then access `https://3tmmtz-3000.csb.app/post/404`, it shows 404 page which is expected.
Run `yarn build`, then `yarn start` in prod mode, access the same url as above, it shows the rendering page, not 404 page.

### Describe the Bug

Leaving dynamicParams=false and accessing an entry point that does not exist in generateStaticParams does not return a 404 page in **prod**. (dev works fine)

The `generateStaticParams` only returns id 1,2,3 which means all the other param with id should return 404. Seems dev works as expected but prod is not.

### Expected Behavior

when access a param that not returned by `generateStaticParams` function, it  should return 404 page in both dev and prod.

**Comment by romainbenard** [comment]
Hello @leozhao0709 did you found a solution?

I have the same issue here with `next 13.5.2`

**Comment by leozhao0709** [comment]
@romainbenard No, the app router approach seems never resolve this issues in production side. I feel the approuter approach is too buggy, so if the app is for production, I used the old way to create next js app.

**Comment by JoiGud** [comment]
We are also experiencing this issue. We tried reproducing it locally by setting `NODE_ENV` to `production` but that still gave us the 404 response; while on our live page we get some weird soup of a 

1.  404 top part / `<header>` and
2.  the main body of the layout is our frontpage

Say, for instance, we request `awesomesite.com/foo` and `/foo` is not a route that is valid, then we get the `404` page. But if we go on deeper; `awesomesite.com/foo/bar` (which also doesn't exist) then we get this weird mixture of layouts and no 404.

We are using Prismic for our CMS, and use their client to generate the static paths.

**Comment by jahirfiquitiva** [comment]
I'm facing the same issue using Next.js `14.0.5-canary.56`

**Comment by azizoid** [comment]
Facing the same problem
`Next.js 14.0.4
   - Environments: .env.local, .env

 ✓ Creating an optimized production build
 ✓ Compiled successfully
 ✓ Linting and checking validity of types
 ✓ Collecting page data
 ✓ Generating static pages (121/121)
 ✓ Collecting build traces
 ✓ Finalizing page optimization`

**Comment by leozhao0709** [comment]
This is still a problem today. 

I found something interesting more today:

when use fetch, if I fetch data within component, then it will work. eg,

```
export function generateStaticParams() {
  return [
    {
      id: '1',
    },
    {
      id: '2',
    }
  ];
}

export default async function PostDetail({
  params,
}: {
  params: { id: string };
}) {
  const { id } = params;

  const post: Post = await (
    await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`)
  ).json();

  return (
    <div>
      This is Post detail:
      {post.id}. {post.body}
    </div>
  );
}

export const dynamicParams = false;
```

This way it will work. But if I extract fetch logic to another function, even the function defined in the same component, then the problem will occur, like the following code, it will not work and show the issue I described in the issue.

```
export function generateStaticParams() {
  return [
    {
      id: '1',
    },
    {
      id: '2',
    }
  ];
}

const getPost = async (id: string) => {
  return await (
    await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
      next: { revalidate: 0 },
    })
  ).json();
};

export default async function PostDetail({
  params,
}: {
  params: { id: string };
}) {
  const { id } = params;

  const post: Post = await getPost(id);

  return (
    <div>
      This is Post detail:
      {post.id}. {post.body}
    </div>
  );
}

export const dynamicParams = false; 
```

This code will show the issue. So I guess the problem is with fetch itself. If we wrap the fetch in a function, then it will not work.

**Comment by leozhao0709** [comment]
One more observation: 

- when using axios instead of fetch to fetch data, then both dev and prod can work correctly. ✅
- When using fetch with `{next: { revalidate: 0 },}, then dev is working, but prod it not work` ❌
- When using fetch **without** `{next: { revalidate: 0 },}`, then both dev and prod can work correctly.  ✅

So I believe the problem is because the nextjs fetch cache. **If you encounter the similar problem, you can either remove revalidate prop(default revalidate is false) or use other http client to fetch data.** 

One thing to remember is based on document here: https://nextjs.org/docs/app/api-reference/functions/fetch#revalidate, when using fetch, if we are not setting the revalidate, the default value is false which means there is a force cache happen.

**Comment by jacekkuczynski** [comment]
I've implemented a workaround, and it works just fine.

 ```
 export async function generateStaticParams() {
  const allCategoriesData = await getAllCategoriesData();
  return allCategoriesData.map((category) => ({
    slug: category.slug,
  }));
}

export default async function Home({ params }: { params: { slug: string } }) {
  const { slug } = params;

  // own implementation of dynamicParams = false
  const allCategoriesData = await getAllCategoriesData();
  const allParams = allCategoriesData.map((category) => category.slug);
  if (!allParams.includes(slug)) notFound();

  const { content } = await getCategoryData(slug);
  return (
    <div className="container flex flex-col items-center justify-center w-full min-h-screen">
      <CategoryContent content={content} />
    </div>
  );
}
```

**Comment by joselinki25** [comment]
Also facing the same problem as of today.

 ```
▲ Next.js 14.1.0
   - Environments: .env.local

   Creating an optimized production build ...
Compiler edge-server unexpectedly exited with code: null and signal: SIGTERM
 ✓ Compiled successfully
 ✓ Linting and checking validity of types    
 ⚠ Using edge runtime on a page currently disables static generation for that page
 ✓ Collecting page data
 ✓ Generating static pages (15/15)
 ✓ Collecting build traces    
 ✓ Finalizing page optimization
```

**Comment by matthewca** [comment]
Also facing the same problem with 14.2.3. I'm trying to limit our website to specific languages (example.com/en and example.com/fr), but I can't get it to consistently throw a 404 for random values.

**Comment by dkottas-nelios** [comment]
> This is still a problem today.
> 
> I found something interesting more today:
> 
> * when using axios instead of fetch to fetch data, then both dev and prod can work correctly. ✅
> * When using fetch with `{next: { revalidate: 0 },}`, then dev is working, but prod it not work ❌
> * When using fetch **without** `{next: { revalidate: 0 },}`, then both dev and prod can work correctly.  ✅
> 
> So I believe the problem is because the nextjs fetch cache. **If you encounter the similar problem, you can either remove revalidate prop(default revalidate is false) or use other http client to fetch data.**
> 
> One thing to remember is based on document here: https://nextjs.org/docs/app/api-reference/functions/fetch#revalidate, when using fetch, if we are not setting the revalidate, the default value is false which means there is a force cache happen.

I am using fetch without {next: { revalidate: 0 },} and still have the same issue in Next 14.2.3.

**Comment by dkottas-nelios** [comment]
For me the problem is with Vercel. I have the exact same project on Vercel and Netlify. Netlify returns 404 as expected, Vercel doesn't.

**Comment by atakanuludag** [comment]
Isn't this problem fixed?

**Comment by hulh122** [comment]
Same with mine, hope to fix the problem.
I have to do like thie in my project, it seems bad

```
export async function generateStaticParams() {
    return await getPaths()
}

export async function generateMetadata({ params }: { params: { id: string } }): Promise<Metadata> {
   const paths = await getBlogPaths()

    if (paths.every(({ id }) => id !== params.id)) {
        return notFound()
    }
}

export default async function Page({ params }: { params: { id: string } }) {
    const paths = await getBlogPaths()

    if (paths.every(({ id }) => id !== params.id)) {
        return notFound()
    }
}
```

**Comment by agustints** [comment]
Hi 👋 , I was wondering if there are any updates on this? I'm encountering the same issue. Thanks in advance!

**Comment by 634750802** [comment]
same in 15.1.4
